ingenialink/ethercat/servo.py

```diff
@@ -150,57 +159,115 @@ def __init__(
-            if e.wkc == self.TIMEOUT_WORKING_COUNTER:
-                raise ILTimeoutError("Timeout reading data.") from e
-            raise ILIOError("Error reading data") from e
+        except (
+            pysoem.SdoError,
+            pysoem.MailboxError,
+            pysoem.PacketError,
+            pysoem.WkcError,
+            ILIOError,
+        ) as e:
+            self._handle_sdo_exception(reg, SDO_OPERATION_MSG.READ, e)
         except pysoem.Emergency as e:
-            error_description = self._get_emergency_description(e.error_code)
-            if error_description is None:
-                error_description = e
-            raise ILIOError(f"Error reading {reg.identifier}. Reason: {error_description}") from e
+            while time.time() < start_time + self._connection_timeout:
+                try:
+                    value = self.__slave.sdo_read(reg.idx, reg.subidx, buffer_size, complete_access)
+                except pysoem.Emergency:
+                    continue
+                except (
+                    pysoem.SdoError,
+                    pysoem.MailboxError,
+                    pysoem.PacketError,
+                    pysoem.WkcError,
+                    ILIOError,
+                ) as exc:
+                    self._handle_sdo_exception(reg, SDO_OPERATION_MSG.READ, exc)
+                else:
+                    return value
+            raise ILTimeoutError("Emergency messages could not be cleared.") from e
         finally:
             self._lock.release()
         return value
 
     def _write_raw(self, reg: EthercatRegister, data: bytes, complete_access: bool = False) -> None:  # type: ignore [override]
         self._lock.acquire()
+        start_time = time.time()
         try:
             self.__slave.sdo_write(reg.idx, reg.subidx, data, complete_access)
-        except (pysoem.SdoError, pysoem.MailboxError, pysoem.PacketError, ILIOError) as e:
-            raise ILIOError(f"Error writing {reg.identifier}. Reason: {e}") from e
-        except pysoem.WkcError as e:
-            if e.wkc == self.NOFRAME_WORKING_COUNTER:
-                raise ILIOError("Error writing data: No frame.") from e
-            if e.wkc == self.TIMEOUT_WORKING_COUNTER:
-                raise ILTimeoutError("Timeout writing data.") from e
-            raise ILIOError("Error writing data") from e
+        except (
+            pysoem.SdoError,
+            pysoem.MailboxError,
+            pysoem.PacketError,
+            pysoem.WkcError,
+            ILIOError,
+        ) as e:
+            self._handle_sdo_exception(reg, SDO_OPERATION_MSG.WRITE, e)
         except pysoem.Emergency as e:
-            error_description = self._get_emergency_description(e.error_code)
-            if error_description is None:
-                error_description = e
-            raise ILIOError(f"Error writing {reg.identifier}. Reason: {error_description}") from e
+            while time.time() < start_time + self._connection_timeout:
```

Comment:
This code is duplicated, did you think about calling the function again and adding an optional parameter with the first-time call time or the number of retries done? You should do that after the lock release

Reply:
Done